tools/eliminate-duplicate-functions.js

```diff
@@ -0,0 +1,524 @@
+      }
+    }
+  });
+}
+
+function replaceDuplicateFuncs(ast, equivalentFunctionHash) {
+  debugLog(ZONE_TOP_LEVEL, "replaceDuplicateFuncs");
+
+  // Replace references to all functions with their equivalent function
+  replaceFunctionReferences(ast, equivalentFunctionHash);
+
+  // Now lets replace the function definitions
+  replaceFunctionDefinitions(ast, equivalentFunctionHash);
+}
+
+function logEquivalentFunctionHash(equivalentFunctionHash) {
+  if (debug && ZONE_EQUIVALENT_FUNCTION_HASH != 0) {
+    debugLog(ZONE_EQUIVALENT_FUNCTION_HASH, "Equivalent Function Hash:");
+    for (var fn in equivalentFunctionHash) {
+      debugLog(ZONE_EQUIVALENT_FUNCTION_HASH, fn + "->" + equivalentFunctionHash[fn]);
+    }
+  }
+}
+
+function generateEquivalentFunctionHash(functionHashToFunctionName, variableNames) {
+  var equivalentFunctionHash = {};
+
+  debugLog(ZONE_TOP_LEVEL, "generateEquivalentFunctionHash");
+
+  if (debug && ZONE_EQUIVALENT_FUNCTION_HASH != 0) {
+    debugLog(ZONE_EQUIVALENT_FUNCTION_HASH, "Equivalent Functions:");
+
+    for (var fnHash in functionHashToFunctionName) {
+      if (functionHashToFunctionName[fnHash].length > 1) {
+        debugLog(ZONE_EQUIVALENT_FUNCTION_HASH, JSON.stringify(functionHashToFunctionName[fnHash], null, '  '));
+      }
+    }
+  }
+
+  for (var fnHash in functionHashToFunctionName) {
+    var equivalentFunctions = functionHashToFunctionName[fnHash];
+    var shortestFunction = undefined;
+    var equivalentFn = undefined;
+
+    // From each list of equivalent functions, pick the
+    // shortest one that is not also a variable name
+    for (var index in equivalentFunctions) {
+      equivalentFn = equivalentFunctions[index];
+
+      // If one of the variables is not the same name as the equivalent function,
+      // and the equivalent function is shorter than the shortest function.
+      if ((variableNames[equivalentFn] === undefined) &&
+        (shortestFunction === undefined || equivalentFn.length < shortestFunction.length)) {
+        shortestFunction = equivalentFn;
+      }
+
+      if (debug && variableNames[equivalentFn] !== undefined) {
+        debugLog(ZONE_EQUIVALENT_FUNCTION_HASH, equivalentFn + " is a variable");
+      }
+    }
+
+    if (shortestFunction !== undefined) {
+      // Populate the equivalent function hash with this info
+      for (var index in equivalentFunctions) {
+        equivalentFn = equivalentFunctions[index];
+
+        // If we're not the shortest function, and
+        // we are not a variable name
+        if ((equivalentFn !== shortestFunction) && variableNames[equivalentFn] === undefined) {
+          equivalentFunctionHash[equivalentFn] = shortestFunction;
+          debugLog(ZONE_EQUIVALENT_FUNCTION_HASH, equivalentFn + "->" + shortestFunction);
+        }
+      }
+    }
+  }
+
+  return equivalentFunctionHash;
+}
+
+function getBodyForFunction(ast, functionName) {
+  var functionBody = '';
+  traverse(ast, function(node, type) {
```

Comment:
this is going to run through the entire AST, including all internal nodes. that's very wasteful. instead, we could scan once and save a mapping. Or, we can just scan through the toplevel nodes that are defun (ast should be a `toplevel`, so at index 1 is a list, and in that list are the defuns. Unless we need to look in an interior scope).